cmd/hack/db/lmdb.go

```diff
@@ -290,6 +291,7 @@ func launchReader(kv ethdb.RwKV, tx ethdb.Tx, expectVal string, startCh chan str
 	}
 	// Wait for the signal to start reading
 	go func() {
+		defer func() { debug.RecoverStackTrace(nil, true, recover()) }()
```

Comment:
If we'd like to report not crush (I'd prefer report and crush immediately), it might be better to introduce a decorator that accepts func() or func() error and runs it in a goroutine with a proper defer?

Reply:
Thanks for the feedback, I will look into it. Can you point me to some places in erigon that use this pattern?

Reply:
I was thinking about something like https://pkg.go.dev/golang.org/x/sync/errgroup
Or https://github.com/lazada/awg
However you can add additional callback to call on panic event.

Reply:
In your top level comment, do you mean doing something like this for the anonymous goroutines? 
 ` func Wrap(f func()) func() {
          return func() {
                  defer RecoverStackTrace(nil, true, recover())
                  f()
          }
 }`

Reply:
```
func Go(toRun func(), onPanic func()) {
    go func() {
        defer func(){
        // handle the panic
            if onPanic != nil {
                onPanic()
            }
        }()
        toRun()
    }()

Reply:
Ah yes thanks. That makes sense for the goroutines that are func() , but most of them aren't. How would one hande situations where toRun is a function like discover(out chan<- *upnp, target string, matcher func(goupnp.ServiceClient) *upnp) or (pool *TxPool) scheduleReorgLoop()

Reply:
A good point. Could it be that it's possible to wrap any function to func() or func error? We could have closures to reduce amount of parameters.

Reply:
Yes, thanks for pushing me down the right track. So RecoverStackTrace would look like this
```
 func Go(toRun func(), onPanic func(interface{}) error) {
          go func() {
                  if onPanic != nil {
                          defer func() { onPanic(recover()) }()
                  }
                  toRun()
          }()
  }
 
 func RecoverStackTrace(err error, stopErigon bool, r interface{}) func(interface{}) error {
          return func(r interface{}) error {
                  if r != nil {
                          fmt.Println("HELLO")
                          stack := string(debug.Stack())
                          switch typed := r.(type) {
                          case error:
                                  err = fmt.Errorf("%w, trace: %s", typed, stack)
                          default:
                                  err = fmt.Errorf("%+v, trace: %s", typed, stack)
                          }
                  }
                  return err
          }
 }
```
So then there is go discover(... in p2p/nat/natupnp.go , it's called something like this:
```
         toRun := func() { 
           discover(found, internetgateway2.URN_WANConnectionDevice_2, func(sc goupnp.ServiceClient) *upnp {
                  switch sc.Service.ServiceType {
                  case internetgateway2.URN_WANIPConnection_1:
                          return &upnp{service: "IGDv2-IP1", client: &internetgateway2.WANIPConnection1{ServiceClient: sc}}
                  case internetgateway2.URN_WANIPConnection_2:
                          return &upnp{service: "IGDv2-IP2", client: &internetgateway2.WANIPConnection2{ServiceClient: sc}}
                  case internetgateway2.URN_WANPPPConnection_1:
                          return &upnp{service: "IGDv2-PPP1", client: &internetgateway2.WANPPPConnection1{ServiceClient: sc}}
                  }
                  return nil
        })
    }
    Go(toRun, RecoverStackTrace(nil, true, recover()))
```

Reply:
@JekaMas 
- I prefer to have current implementation rather than wrapper, because I don’t see difference: wrapper only hiding something - but why need hiding something from me? 

- About “why mot fail fast”: imagine that your business is based on serving RPC calls - and you don’t really care if node is in sync or not. And Erigon on some new block got nil-pointer panic - fail fast means here that you will not able to start serving RPC requests for very long time - despite that RPC functionality working great just node can’t sync.

Reply:
@bitbasebit my PR’s never getting so much comments. You are popular :-)

Reply:
1. I just were thinking about a single way of doing the thing with "not forget to add a right defer property". For me it's better to hide code duplication.
2. Discussable. On the other hand we do have cases when going further after panic could lead to incorrect RPC handling or data corruption. I think it's just a team decision: what case is more important. I am ok with any.